### ISSUE TITLE
the transform of grasp pose

### ISSUE THREAD

**Issue by MengDes** [dougsm/ggcnn_kinova_grasping#4]
hi, i am reading the paper and code about ggcnn
but i find something inconsistent with my thought
in kinova_closed_loop_grasp.py
[https://github.com/dougsm/ggcnn_kinova_grasping/blob/master/ggcnn_kinova_grasping/scripts/kinova_closed_loop_grasp.py#L100](url)

-  
           # Construct the Pose in the frame of the camera.
            gp = geometry_msgs.msg.Pose()
            gp.position.x = d[0]
            gp.position.y = d[1]
            gp.position.z = d[2]
            q = tft.quaternion_from_euler(0, 0, -1 * d[3])
            gp.orientation.x = q[0]
            gp.orientation.y = q[1]
            gp.orientation.z = q[2]
            gp.orientation.w = q[3]

why the yaw angle is -1*d[3]? （i think φ should be counterclockwise rotation around z axis ,so it should be d[3],and  kinova_open_loop_grasp.py,，https://github.com/dougsm/ggcnn_kinova_grasping/blob/master/ggcnn_kinova_grasping/scripts/kinova_open_loop_grasp.py#L77also did not multiply -1 ）

       # Average pose in base frame.
        gp_base.position.x = av[0]
        gp_base.position.y = av[1]
        gp_base.position.z = av[2]
        GOAL_Z = av[2]
        ang = av[3] - np.pi/2  # We don't want to align, we want to grip.
        q = tft.quaternion_from_euler(np.pi, 0, ang)
        gp_base.orientation.x = q[0]
        gp_base.orientation.y = q[1]
        gp_base.orientation.z = q[2]
        gp_base.orientation.w = q[3]
 
i am quite confues with **ang = av[3] - np.pi/2** ，the meaning to  minus -pi/2
here is fig in the paper
![image](https://user-images.githubusercontent.com/27760271/54686588-9e7f7500-4b54-11e9-809d-9132cf6f7893.png)
i think i made some conceptual mistaek,so i'll be gratefull if you do some explanation about my question

@dougsm

**Comment by dougsm** [comment]
Hello,
Sorry that's poorly documented - I think both are just artefacts of the network implementation and its frame of reference.
* Q1: The -1*d[3] is just inverting the rotation direction. In the training data, the angle is computed as the required rotation of the image, but in the frame of reference of the camera this rotation is reversed. That is, a clockwise rotation of the image corresponds to a counter-clockwise rotation of the camera and vice versa.
* Q2: Similarly, the angle output from the network is actually aligned along the axis of the grasp, so in order to actually grasp the object we want the gripper to be perpendicular to this axis so subtract pi/2.

**Comment by MengDes** [comment]
got it. i should figure out the dataset first
thank you

**Comment by dougsm** [comment]
Actually, I just realised my answer to Q2 is wrong.
The reason is that the end effector frame, which isn't pictured but is the frame that the frame that is controlled, is actually 90 degrees out from the camera frame.  The robot end effector x-axis is aligned with the camera y-axis.
